### PR TITLE
Continuation of “Weather.com smartspace provider”

### PR DESCRIPTION
**Pull request by ghost**
#1661 was closed due to being nothing more than a blind cherry-pick

This hopefully fixes most of the pulled-in librechair code

### CODE REVIEW

File: lawnchair/src/ch/deletescape/lawnchair/preferences/OWMEditTextPreference.kt [hunk 1]
```diff
@@ -50,10 +49,7 @@ class OWMEditTextPreference(context: Context, attrs: AttributeSet?) : EditTextPr
     }
 
     override fun onValueChanged(key: String, prefs: LawnchairPreferences, force: Boolean) {
-        if (key == "pref_smartspace_widget_provider") {
-            isVisible = prefs.weatherProvider == OWMWeatherDataProvider::class.java.name
-                    || !isApi && prefs.weatherProvider == AccuWeatherDataProvider::class.java.name
-        }
+        isVisible = !(isApi && prefs.weatherProvider != OWMWeatherDataProvider::class.java.name)
```
Comment (nyancrimew) [inline lawnchair/src/ch/deletescape/lawnchair/preferences/OWMEditTextPreference.kt:52]: This breaks city selection for the Accuweather Provider

File: lawnchair/src/ch/deletescape/lawnchair/preferences/SmartspaceProviderPreference.kt [hunk 2]
```diff
@@ -58,17 +58,19 @@ class SmartspaceProviderPreference(context: Context, attrs: AttributeSet?)
...
         list.add(AccuWeatherDataProvider::class.java.name)
         list.add(OWMWeatherDataProvider::class.java.name)
+        list.add(WeatherChannelWeatherProvider::class.java.name)
         if (PEWeatherDataProvider.isAvailable(context))
             list.add(PEWeatherDataProvider::class.java.name)
         if (OnePlusWeatherDataProvider.isAvailable(context))
             list.add(OnePlusWeatherDataProvider::class.java.name)
         if (prefs.showDebugInfo)
             list.add(FakeDataProvider::class.java.name)
+        if (prefs.showDebugInfo) list.add(FakeDataProvider::class.java.name)
```
Comment (nyancrimew) [inline lawnchair/src/ch/deletescape/lawnchair/preferences/SmartspaceProviderPreference.kt]: This is already being done on the line above

File: lawnchair/src/ch/deletescape/lawnchair/smartspace/LawnchairSmartspaceController.kt [hunk 3]
```diff
@@ -523,7 +527,8 @@ class LawnchairSmartspaceController(val context: Context) {
                 Pair(BatteryStatusProvider::class.java.name, R.string.battery_status),
                 Pair(PersonalityProvider::class.java.name, R.string.personality_provider),
                 Pair(OnboardingProvider::class.java.name, R.string.onbording),
-                Pair(FakeDataProvider::class.java.name, R.string.weather_provider_testing))
+                Pair(FakeDataProvider::class.java.name, R.string.weather_provider_testing),
+                WeatherChannelWeatherProvider::class.java.name to R.string.title_weather_provider_weather_com)
```
Comment (nyancrimew) [inline lawnchair/src/ch/deletescape/lawnchair/smartspace/LawnchairSmartspaceController.kt]: Either adhere to the existing code style or change everything.

File: lawnchair/src/ch/deletescape/lawnchair/smartspace/weathercom/Constants.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+package ch.deletescape.lawnchair.smartspace.weathercom
+
+import ch.deletescape.lawnchair.smartspace.WeatherIconProvider
+
+
+object Constants {
+
+    object WeatherComConstants {
+        val WEATHER_COM_API_KEY = "8de2d8b3a93542c9a2d8b3a935a2c909"
```
Comment (nyancrimew) [inline lawnchair/src/ch/deletescape/lawnchair/smartspace/weathercom/Constants.kt]: Please extract into build time variable just like with accuweather

File: lawnchair/src/ch/deletescape/lawnchair/smartspace/weather/weathercom/WeatherChannelWeatherProvider.kt [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+class WeatherChannelWeatherProvider(controller: LawnchairSmartspaceController) :
+        PeriodicDataProvider(controller) {
+
+    override fun updateData() {
+        runOnUiWorkerThread {
+            if (context.lawnchairPrefs.weatherCity != "##Auto") {
+                try {
+                    val position = WeatherComRetrofitServiceFactory.weatherComWeatherRetrofitService
+                            .searchLocationByName(context.lawnchairPrefs.weatherCity, "city",
+                                                  context.locale.language, "json").execute()
```
Comment (nyancrimew) [inline lawnchair/src/ch/deletescape/lawnchair/smartspace/weather/weathercom/WeatherChannelWeatherProvider.kt]: If you already abstracted everything, including the JSON decoding out why leave the format up to the caller? This method definitely doesn't work with anything else than "json" being supplied. Same with the "city" parameter, change the method to `searchCityByName` and have city already specified in the service.

File: lawnchair/src/ch/deletescape/lawnchair/smartspace/weathercom/Constants.kt [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+                                          39 to WeatherIconProvider.CONDITION_SHOWERS,
+                                          40 to WeatherIconProvider.CONDITION_RAIN,
+                                          41 to WeatherIconProvider.CONDITION_RAIN,
+                                          42 to WeatherIconProvider.CONDITION_SNOW,
+                                          43 to WeatherIconProvider.CONDITION_SNOW,
+                                          44 to WeatherIconProvider.CONDITION_UNKNOWN,
+                                          45 to WeatherIconProvider.CONDITION_SHOWERS,
+                                          46 to WeatherIconProvider.CONDITION_SNOW,
+                                          47 to WeatherIconProvider.CONDITION_STORM)
+        val WEATHER_ICONS_DAY = WEATHER_ICONS.map { it.key to it.value + "d" }
```
Comment (nyancrimew) [inline lawnchair/src/ch/deletescape/lawnchair/smartspace/weathercom/Constants.kt]: Just removing "d" and "n" from the icon code at the call site would probably work as well as the API already supplies us with day/night info

File: res/values/strings.xml [hunk 7]
```diff
@@ -51,9 +51,9 @@
     <string name="widget_dims_format">%1$d \u00d7 %2$d</string>
     <!-- Accessibility spoken message format for the dimensions of a widget in the drawer -->
     <string name="widget_accessible_dims_format">%1$d wide by %2$d high</string>
-    <!-- Message to tell the user to press and hold a widget/icon to add it  -->
+    <!-- Message to tell the user to press and hold a widget/iconView to add it  -->
```
Comment (nyancrimew) [inline res/values/strings.xml]: ?

File: res/values/strings.xml [hunk 8]
```diff
@@ -51,9 +51,9 @@
...
     <string name="add_item_request_drag_hint">Touch &amp; hold to place manually</string>
-    <!-- Button label to automatically add icon on home screen [CHAR_LIMIT=50] -->
+    <!-- Button label to automatically add iconView on home screen [CHAR_LIMIT=50] -->
```
Comment (nyancrimew) [inline res/values/strings.xml]: ?

File: src/com/android/launcher3/Utilities.java [hunk 9]
```diff
@@ -882,6 +879,8 @@ public static void openURLinBrowser(Context context, String url, Rect sourceBoun
         } catch (ActivityNotFoundException exc) {
             // Believe me, this actually happens.
             Toast.makeText(context, R.string.error_no_browser, Toast.LENGTH_SHORT).show();
+        } catch (NullPointerException e) {
```
Comment (nyancrimew) [inline src/com/android/launcher3/Utilities.java]: Instead of catching an NPE here check the URL passed in at the beginning using TextUtils

File: src/com/android/launcher3/Utilities.java [hunk 10]
```diff
@@ -159,7 +156,7 @@
      */
     public static final boolean IS_DEBUG_DEVICE =
             Build.TYPE.toLowerCase(Locale.ROOT).contains("debug") ||
-            Build.TYPE.toLowerCase(Locale.ROOT).equals("eng");
+                    Build.TYPE.toLowerCase(Locale.ROOT).equals("eng");
```
Comment (nyancrimew) [inline src/com/android/launcher3/Utilities.java:159]: Can we get rid of all formatting changes in unrelated code in this PR?

File: lawnchair/src/ch/deletescape/lawnchair/smartspace/weathercom/models/SunV1CurrentConditionsResponse.java [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+        public Integer minTemp;
+
+        @SerializedName("obs_id")
+        public String obsId;
+
+        @SerializedName("obs_name")
+        public String obsName;
+
+        @SerializedName("precip_hrly")
+        public Integer precipHrly;
```
Comment (nyancrimew) [inline lawnchair/src/ch/deletescape/lawnchair/smartspace/weathercom/models/SunV1CurrentConditionsResponse.java]: As you are already using SerializedName here I suggest you to give all variables even more readable names (just expanding abbreviations should generally be enough)

**Review by nyancrimew — CHANGES_REQUESTED** [review]

File: lawnchair/src/ch/deletescape/lawnchair/smartspace/weather/weathercom/WeatherChannelWeatherProvider.kt (hunk 5 above)
Comment (ghost) [reply lawnchair/src/ch/deletescape/lawnchair/smartspace/weather/weathercom/WeatherChannelWeatherProvider.kt]: The problem is that the Weather.com API doesn’t always allow a “json” parameter, so specifying it in the service would not exactly be ideal

Also, I plan on eventually adding alternative search methods, so I actually think this is more logical

File: lawnchair/src/ch/deletescape/lawnchair/smartspace/weathercom/Constants.kt (hunk 4 above)
Comment (ghost) [reply lawnchair/src/ch/deletescape/lawnchair/smartspace/weathercom/Constants.kt]: OK

File: lawnchair/src/ch/deletescape/lawnchair/smartspace/LawnchairSmartspaceController.kt (hunk 3 above)
Comment (ghost) [reply lawnchair/src/ch/deletescape/lawnchair/smartspace/LawnchairSmartspaceController.kt]: I’ll push a change later once I get off the train

File: lawnchair/src/ch/deletescape/lawnchair/preferences/SmartspaceProviderPreference.kt (hunk 2 above)
Comment (ghost) [reply lawnchair/src/ch/deletescape/lawnchair/preferences/SmartspaceProviderPreference.kt]: Hmm interesting

File: lawnchair/src/ch/deletescape/lawnchair/preferences/OWMEditTextPreference.kt (hunk 1 above)
Comment (ghost) [reply lawnchair/src/ch/deletescape/lawnchair/preferences/OWMEditTextPreference.kt:52]: The city selection works for me though

File: lawnchair/src/ch/deletescape/lawnchair/preferences/SmartspaceProviderPreference.kt (hunk 2 above)
Comment (ghost) [reply lawnchair/src/ch/deletescape/lawnchair/preferences/SmartspaceProviderPreference.kt]: I’ll push a change

File: lawnchair/src/ch/deletescape/lawnchair/smartspace/weathercom/models/SunV1CurrentConditionsResponse.java (hunk 11 above)
Comment (ghost) [reply lawnchair/src/ch/deletescape/lawnchair/smartspace/weathercom/models/SunV1CurrentConditionsResponse.java]: It’s actually something I was working on, but I forgot about it a couple of hours ago

File: src/com/android/launcher3/Utilities.java (hunk 9 above)
Comment (ghost) [reply src/com/android/launcher3/Utilities.java]: Done

File: src/com/android/launcher3/Utilities.java (hunk 10 above)
Comment (ghost) [reply src/com/android/launcher3/Utilities.java:159]: OK! What I remember is that I actually copy & pasted the official Lawnchair Utilities.java here but Android Studio decided to correct the changes all the same

File: res/values/strings.xml (hunk 7 above)
Comment (ghost) [reply res/values/strings.xml]: Kotlin refactoring is buggy on 3.5

File: lawnchair/src/ch/deletescape/lawnchair/smartspace/weathercom/Constants.kt (hunk 6 above)
Comment (oldosfan) [reply lawnchair/src/ch/deletescape/lawnchair/smartspace/weathercom/Constants.kt]: I’ll take a look at that, although it is due to the fact that the API sometimes doesn’t adhere to API guidelines

File: lawnchair/src/ch/deletescape/lawnchair/smartspace/weather/weathercom/WeatherChannelWeatherProvider.kt (hunk 5 above)
Comment (ghost) [reply lawnchair/src/ch/deletescape/lawnchair/smartspace/weather/weathercom/WeatherChannelWeatherProvider.kt]: I’ve abstracted the format but not type

File: res/values/strings.xml (hunk 8 above)
Comment (ghost) [reply res/values/strings.xml]: A refactor gone wrong

File: src/com/android/launcher3/Utilities.java (hunk 10 above)
Comment (suphon-t) [reply src/com/android/launcher3/Utilities.java:159]: Well you can just discard them from git

Comment (ghost) [reply src/com/android/launcher3/Utilities.java:159]: > Well you can just discard them from git

Which is what I would have done, except that it’s broken in AS 3.5

Anyway I ended up doing it from the command-line

**Comment by ghost** [comment]
@deletescape All requested changes have been resolved.
